### ISSUE TITLE
Is ResNet 3f causal？A question about causal convolution.

### ISSUE THREAD

**Issue by 20050710212** [anicolson/DeepXi#23]
Hi, thanks for the code sharing.

In the **def CausalDilatedConv1d**, **Zeros** are concatenated on the left of **x** to enable dilated convolution. But in the convolution,  data on the right side of the current frame, that is the future data,  are used.  I think only the data on the left of the current frame, which means history data, can be convolved.  Can you help me to understand how the causal convolution is achieved here.

Thank you！
```
def CausalDilatedConv1d(x, d_f, k_size, d_rate=1, use_bias=True):
    if k_size > 1: # padding for causality.
        x_shape = tf.shape(x)
	x = tf.concat([tf.zeros([x_shape[0], (k_size - 1)*d_rate, x_shape[2]]), x], 1)
    return tf.layers.conv1d(x, d_f, k_size, dilation_rate=d_rate, activation=None, padding='valid', use_bias=use_bias)
```

**Comment by anicolson** [comment]
The combination zero padding before the sequence and the use of padding=‘valid’ makes this possible.

Hope this helps.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by anicolson** [comment]
here is a demo with an input size of 1 and a filter size of 1.

![Photo 12-3-20, 10 28 02 am](https://user-images.githubusercontent.com/26111230/76475953-394f1e80-644c-11ea-9b97-83cad24ab034.jpg)

**Comment by 20050710212** [comment]
Thanks, got it.
